### PR TITLE
Mldb 1058 one thread per loop

### PR DESCRIPTION
**Pull request by wsourdeau**
This commit completely isolates event loops from one another, which slightly enhances the performance during the handling of events.

### CODE REVIEW

File: http/asio_thread_pool.cc [hunk 1]
```diff
@@ -27,180 +28,47 @@ namespace Datacratic {
...
 
     ~Impl()
     {
-        shutdown = true;
-        ioService.stop();
-        work.reset();
-        for (auto & t: threads) {
-            ioService.post([] () {});
-            t.join();
+        for (size_t i = 0; i < threads.size(); i++) {
```
Comment (FinchPowers) [inline http/asio_thread_pool.cc]: Tu sais que ++i est plus rapide que i++? Probablement optimisé par le compilateur.

File: http/asio_thread_pool.cc [hunk 2]
```diff
@@ -9,6 +9,7 @@
 
 #include "asio_thread_pool.h"
 #include "asio_timer.h"
+#include "event_loop.h"
```
Comment (FinchPowers) [inline http/asio_thread_pool.cc]: asio_timer.h, <boost/asio.hpp>, watch_impl.h, ne servent pas je crois.

Comment (FinchPowers) [inline http/asio_thread_pool.cc]: Si event_loop_impl.h je crois que event_loop.h est inutile.

File: rest/http_rest_service.cc [hunk 3]
```diff
@@ -272,12 +268,21 @@ init()
                const std::string & payload)
         {
             std::string requestId = this->getHttpRequestId();
-            HttpRestConnection restConnection(connection, requestId, this);
+            HttpRestConnection restConnection(connection,
+                                              std::move(requestId), this);
```
Comment (FinchPowers) [inline rest/http_rest_service.cc]: Pas besoin de std::move si this->getHttpRequestId() est passé directement. Maintenant je ne sais pas si l'optimisation de la compilation fait la substitution automatiquement.

File: http/asio_thread_pool.cc [hunk 4]
```diff
@@ -27,180 +28,47 @@ namespace Datacratic {
...
     {
-        shutdown = true;
-        ioService.stop();
-        work.reset();
-        for (auto & t: threads) {
-            ioService.post([] () {});
-            t.join();
+        for (size_t i = 0; i < threads.size(); i++) {
+            loops[i]->terminate();
+            threads[i].join();
```
Comment (FinchPowers) [inline http/asio_thread_pool.cc]: Si je ne me trompe pas ça ne devrait pas être ici. thread.join n'a pas de "nothrow" alors ton dtor peut thrower. Not good.
http://en.cppreference.com/w/cpp/thread/thread/join

File: http/asio_thread_pool.cc (hunk 1 above)
Comment (wsourdeau) [inline http/asio_thread_pool.cc]: Non. Ça me surprend.

File: http/asio_thread_pool.cc (hunk 2 above)
Comment (wsourdeau) [inline http/asio_thread_pool.cc]: Tu as raison pour les 3 premiers, je dois vérifier pour les autres.

File: rest/http_rest_service.cc (hunk 3 above)
Comment (wsourdeau) [inline rest/http_rest_service.cc]: Non, le compilateur ne le fait pas. requestId est et reste une lvalue.

File: http/asio_thread_pool.cc (hunk 4 above)
Comment (wsourdeau) [inline http/asio_thread_pool.cc]: Techniquement, c'est vrai. Je vais mettre un try/catch {abort()}

**Comment by nicolaskruchten** [comment]
what's the status of this PR?

**Comment by wsourdeau** [comment]
@nicolaskruchten : I need to illustrate the performance gain with a benchmark.

**Comment by wsourdeau** [comment]
@nicolaskruchten @jeremybarnes gain = 10kqps too

**Comment by jeremybarnes** [comment]
I'm reluctant, as it looks like we're optimizing towards a benchmark and I can't convince myself that it's not going to harm general purpose workloads.  Can you further explain the rationale, implemention and potential impact, as well as split into two commits (one which adds the ensureThreads functionality but doesn't impact the settings and thus has zero performance impact, and one which actually modifies the settings to make the benchmark faster)?

**Comment by wsourdeau** [comment]
@jeremybarnes I will

**Comment by jeremybarnes** [comment]
Closing for now, until the next round of benchmarking.  Leaving the branch there, though.

**Pull request closed without merging**